Removed the  line where the directory is stripped off the RelativeFile

MarcelEdward opened the pull request:
Removed the line where the directory is stripped off the RelativeFilename, and replace \ with /.

However this introduces a new problem with the current fbx files who do not have their textures in the original dir but moved it to the dir where the fbx file lives.

Suppose there should be some logic when the texture is not found to look if the texture lives in the dir above the RelativeFilename.
In any case it is not obvious when there is a texture missing on a model, it simply renders black.

hifi-gustavo commented:
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

cozza13 commented:
ok to test

cozza13 commented:
this relates to this worklist job: https://worklist.net/20605

ZappoMan commented:
What's the testing plan on this? Have we verified that this change doesn't break any content?

murillodigital commented:
ok to test

hifi-gustavo commented:
Build console output for commit: fac497dadc139f3c1e649178c1ccf2fc2c0e4e14 available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg

MarcelEdward commented:
The build log does not seem to make sense. I did not do anything with ErrorDialog.cpp

And yes it will 'break' content.
The fbx expects to find the textures at the place in the directory specified in the RelativeFile constant out of the fbx itself.
It is not logical to put the textures in the same dir and overrule the constants in the fbx itself. That infact 'breaks' the content since it will no longer display correct when downloaded and edited a 3D editor.

Do you want some logic in HF when textures are not found ?

ZappoMan commented:
test this please

hifi-gustavo commented:
Build console output for commit: fac497dadc139f3c1e649178c1ccf2fc2c0e4e14 available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg

hifi-gustavo commented:
Build console output for commit: 8f0893ba2153b5a10f22aa0cff4cd573de3063cf available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg

MarcelEdward commented:
Added fileOnUrl to check if a texture exist at the location. It retuns the correct filename of where the texture lives.
Added the url of the fix file to extractFBXGeometry and readFBX and updated the calls to readFBX to include the url of the fix file.

So it now does not break existing content.

Found a second place in the FBXReader.cpp where the RelativeFileName stripped out the dir location.

cozza13 commented:
test this please

hifi-gustavo commented:
Build console output for commit: 8f0893ba2153b5a10f22aa0cff4cd573de3063cf available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg

cozza13 commented:
@ZappoMan I tested this on OSX and checked out old content

AndrewMeadows commented on libraries/fbx/src/FBXReader.cpp:
```diff
@@ -1454,9 +1455,23 @@ void buildModelMesh(ExtractedMesh& extracted) {
+    filename = filename.mid(qMax(filename.lastIndexOf('\\'), filename.lastIndexOf('/')) + 1);
+    filename = filename.replace('\\', '/');
+    return filename;
+}
```
(1) Is the **fileOnUrl** function name a typo?  Was it supposed to be **fileOrUrl**?

(2) I was having a little bit of trouble parsing the logic.  It appears that the function is sometimes trying to extract a filename out of an URL that has delimiters that are... slashes in the wrong direction?  These strange characters... where do they come from?  Are they a Windows-only thing (if so: Damn you MS for your everlasting curse of bad path delimiters!)? or is it a an FBX thing?

The final value of filename... it is the string between the last **\** and the character after the last **/**... and then the 

In the case where **checkfile** is a valid file it should be possible to pull the filename directly out of the QFileInfo class rather than later do the special string manipulations (e.g. filename = checkfile.fileName()).  Then we could put all the special string manipulations inside the else case -- clearer logic flow IMO.

MarcelEdward commented:
The function checks if the file exists at an url. In most cases it will look for a texture, but the function can also check for a video. (a video on a model, that is cool. I did not know that was possible :))

In the case of the example fbx from the worklist a texture is expected at External\OutofHere.jpg, the location of the texture is stored inside the fbx file.

So after uploading the saved fbx to somewhere it will look like:

external.fbx
External/OutOfHere.jpg

For some reason the original code stripped out the dir, effectively making the texture unfindable.
The current workaround is to copy the texture to the dir where the fbx lives, obvious this should not be the default thing to do. But I think it is ok to also look for the texture in the dir where the fbx lives when not found at the expected location. This will not break existing content like fbx files stored locally by interface.

We do not want to pull just the filename but also preserve the dir info. So we cannot just pull the filename out of it. 

The \ is a windows thing, it has to be replaced by a /. I do not know why the fbx format is designed to include the windows bad path delimiters. I doubt they even made a decision to do that. But hence that is in stone, and all url's use /.

The last two filename lines should not be there, pushed a patch to remove them.

hifi-gustavo commented:
Build console output for commit: d25e09af74544db662a2e92637439386b00cb284 available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg

PhilipRosedale commented:
@cozza13 @ZappoMan Is this OK to merge?

cozza13 commented:
@PhilipRosedale yes I have tested it.

ZappoMan commented on libraries/fbx/src/FBXReader.cpp:
```diff
@@ -2710,12 +2722,12 @@ FBXGeometry extractFBXGeometry(const FBXNode& node, const QVariantHash& mapping,
 }
 
-FBXGeometry readFBX(const QByteArray& model, const QVariantHash& mapping, bool loadLightmaps, float lightmapLevel) {
+FBXGeometry readFBX(const QByteArray& model, const QVariantHash& mapping, const QString url, bool loadLightmaps, float lightmapLevel) {
```
this should be "const QString& url"

ZappoMan commented on libraries/fbx/src/FBXReader.cpp:
```diff
@@ -2710,12 +2722,12 @@ FBXGeometry extractFBXGeometry(const FBXNode& node, const QVariantHash& mapping,
 
-FBXGeometry readFBX(QIODevice* device, const QVariantHash& mapping, bool loadLightmaps, float lightmapLevel) {
-    return extractFBXGeometry(parseFBX(device), mapping, loadLightmaps, lightmapLevel);
+FBXGeometry readFBX(QIODevice* device, const QVariantHash& mapping, const QString url, bool loadLightmaps, float lightmapLevel) {
```
this should be "const QString& url"

ZappoMan commented on libraries/fbx/src/FBXReader.h:
```diff
@@ -268,10 +268,10 @@ Q_DECLARE_METATYPE(FBXGeometry)
 /// Reads FBX geometry from the supplied model and mapping data.
 /// \exception QString if an error occurs in parsing
-FBXGeometry readFBX(const QByteArray& model, const QVariantHash& mapping, bool loadLightmaps = true, float lightmapLevel = 1.0f);
+FBXGeometry readFBX(const QByteArray& model, const QVariantHash& mapping, const QString url = "", bool loadLightmaps = true, float lightmapLevel = 1.0f);
```
this should be "const QString& url"

ZappoMan commented on libraries/fbx/src/FBXReader.cpp:
```diff
@@ -1454,9 +1455,21 @@ void buildModelMesh(ExtractedMesh& extracted) {
 
-
-FBXGeometry extractFBXGeometry(const FBXNode& node, const QVariantHash& mapping, bool loadLightmaps, float lightmapLevel) {
+FBXGeometry extractFBXGeometry(const FBXNode& node, const QVariantHash& mapping, QString url, bool loadLightmaps, float lightmapLevel) {
```
this should be "const QString& url"

hifi-gustavo commented:
Build console output for commit: fa6d6a51230de1c91b8d2bf0ff29ddc5fe757603 available at: https://gist.github.com/4e0d49d4d9a4f9bf446c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5323.dmg
